### PR TITLE
ILLUSIONS: Fix wide text handling logic for big endian systems. (#11236)

### PR DESCRIPTION
**Pull request by yuv422**
This should fix the issue with wide text processing on big endian systems.

Could someone try this on a big endian system for me to see if I've fixed it.

### CODE REVIEW

**Comment by bgK** [comment]
Hi,

I've tried this branch on a PS3, with the demo version. I believe the game goes a bit further, however it still eventually gets stuck in an infinite loop. This time at the pool scene, after a bit of dialog (right before getting control?).

File: engines/illusions/illusions.cpp [hunk 1]
```diff
@@ -61,19 +61,27 @@
...
-		wstr += 2;
+		*p++ = (byte)*wstr;
+		wstr++;
 		i++;
 	}
 	*p = 0;
 	return buf;
 }
 
+void swapBytesInWideString(byte * wstr) {
```
Comment (bluegr) [inline engines/illusions/illusions.cpp]: Formatting (byte *wstr)

File: engines/illusions/resources/talkresource.cpp [hunk 2]
```diff
@@ -45,11 +45,18 @@ void TalkEntry::load(byte *dataStart, Common::SeekableReadStream &stream) {
...
 	_tblPtr = dataStart + tblOffs;
 	_voiceName = dataStart + voiceNameOffs;
 	debug(0, "TalkEntry::load() _talkId: %08X; textOffs: %08X; tblOffs: %08X; voiceNameOffs: %08X",
 		_talkId, textOffs, tblOffs, voiceNameOffs);
+
+#if defined(SCUMM_BIG_ENDIAN)
+	for (byte *ptr = (byte *)_text; ptr != _tblPtr; ptr += 2) {
+		WRITE_UINT16(ptr, SWAP_BYTES_16(READ_UINT16(ptr)));
+	}
+#endif
```
Comment (bluegr) [inline engines/illusions/resources/talkresource.cpp]: Why don't you use swapBytesInWideString() here?

**Comment by sev-** [comment]
Any update on this?

**Comment by yuv422** [comment]
I got a bit stuck trying to get a bigendian environment setup. I'll take another look this weekend.

**Comment by sev-** [comment]
No news?

**Comment by yuv422** [comment]
I still haven't tested this on a big endian system. I think it is safe to merge though. I'll just check it against master first.

File: engines/illusions/resources/talkresource.cpp (hunk 2 above)
Comment (yuv422) [reply engines/illusions/resources/talkresource.cpp:58]: This one doesn't check for the null terminator

**Comment by raziel-** [comment]
@yuv422

Sorry, but I've  just seen this PR.
I have a BE machine (AmigaOS4) and could do some testing.

Which games use the illusion engine?
Is that Duckman?
If it is then there's already a bug item in the tracker from me.

If not, what games and what exactly would you need me to test?

I can start testing at the end of this week

**Comment by yuv422** [comment]
Hi @raziel- 
Thanks that would be great.
There is an issue in the tracker. https://bugs.scummvm.org/ticket/11236
The game is Duckman. You can use the demo version to test.

**Comment by raziel-** [comment]
@yuv422

I can test in a few days, but I think I already found a duplicate or at least something similar.
See here:
https://bugs.scummvm.org/ticket/11528

**Comment by raziel-** [comment]
@yuv422
https://bugs.scummvm.org/ticket/11236#comment:1

Crash on BE AmigaOS4, crashlog attached to the bug item

**Comment by sev-** [comment]
@yuv422 do you have time to take a look into this?

**Comment by yuv422** [comment]
Ok I tested this against master and it looks ok. I still need to do a bit more to support big endian systems but I'll do that in another PR

**Pull request merged**